src/librustc_typeck/check/wfcheck.rs

```diff
@@ -253,16 +253,55 @@ impl<'ccx, 'gcx> CheckTypeWellFormedVisitor<'ccx, 'gcx> {
         });
     }
 
+    fn check_auto_trait(&mut self,
+                                   trait_def_id: DefId,
+                                   span: Span)
+    {
+        let predicates = self.tcx().lookup_predicates(trait_def_id);
+
+        // If we must exclude the Self : Trait predicate contained by all
+        // traits.
+        let no_refl_predicates : Vec<_> =
```

Comment:
Nit: if the only thing we care about is whether this is empty, perhaps

```rust
predicates.predicates.iter()
   .any(|predicate| ...)
```

However, I think we also want to test that the "self type" is as expected -- this seems to only test the trait. Try adding a test for something like `trait Foo : Option<Self>: Foo`, and I expect it will still be permitted.